boxsdk/object/upload_session.py

```diff
@@ -0,0 +1,146 @@
+
+        :rtype:
+            `unicode`
+        """
+        return self.session.get_url(
+            '{0}s/{1}s'.format(self._parent_item_type, self._item_type),
+            self._object_id,
+            *args
+        ).replace(API.BASE_API_URL, API.UPLOAD_URL)
+
+    def get_parts(self):
+        """
+        Get a list of parts uploaded so far.
+
+        :returns:
+            Returns a `list` of parts uploaded so far.
+        :rtype:
+            `list` of `dict`
+        """
+        response = self.session.get(self.get_url('parts')).json()
+        return response['entries']
+
+    def _calculate_part_sha1(self, content_stream):
+        """
+        Calculate the SHA1 hash of the chunk stream for a given part of the file.
+
+        :returns:
+            The unencoded SHA1 hash of the part.
+        :rtype:
+            `bytes`
+        """
+        content_sha1 = hashlib.sha1()
+        stream_position = content_stream.tell()
+        hashed_length = 0
+        while hashed_length < self.part_size:  # pylint:disable=no-member
+            chunk = content_stream.read(self.part_size - hashed_length)  # pylint:disable=no-member
+            if chunk is None:
+                continue
+            if not chunk:
+                break
+            hashed_length += len(chunk)
+            content_sha1.update(chunk)
+        content_stream.seek(stream_position)
+        return content_sha1.digest()
+
+    def upload_part(self, content_stream, offset, total_size, part_content_sha1=None):
+        """
+        Upload a part of a file.
+
+        :param content_stream:
+            File-like object containing the content of the part to be uploaded.
+        :type content_stream:
+            :class:`File`
+        :param offset:
+            Offset, in number of bytes, of the part compared to the beginning of the file.
+        :type offset:
+            `int`
+        :param total_size:
+            The size of the file that this part belongs to.
+        :type total_size:
+            `int`
+        :param part_content_sha1:
+            SHA-1 hash of the part's content. If not specified, this will be calculated.
+        :type part_content_sha1:
+            `unicode`
+        :returns:
+            The uploaded part.
+        :rtype:
+            `dict`
+        """
+        if part_content_sha1 is None:
+            part_content_sha1 = self._calculate_part_sha1(content_stream)
+
+        range_end = min(offset + self.part_size - 1, total_size - 1)  # pylint:disable=no-member
+
+        return self._session.put(
+            self.get_url(),
+            headers={
+                'Content-Type': 'application/octet-stream',
+                'Digest': 'SHA={0}'.format(base64.b64encode(part_content_sha1).decode('utf-8')),
+                'Content-Range': 'bytes {0}-{1}/{2}'.format(offset, range_end, total_size),
+            },
+            data=content_stream
+        ).json()
+
+    def commit(self, parts, content_sha1):
```

Comment:
This should also allow passing arbitrary file attributes and an ETag; see https://developer.box.com/v2.0/reference/#commit-upload

Also, `parts` can be optional — if the user doesn't pass it in you can just call the API to get the list of parts the server has and use those (if you just want to trust the server).